tests/kvm.rs

```diff
@@ -0,0 +1,201 @@
+mod config;
+
+use env_logger;
+use log::debug;
+use std::mem;
+use std::panic;
+use std::process::{Command, Stdio};
+use std::sync::{mpsc, Once};
+use std::thread;
+use std::time::Duration;
+
+use config::{KVMI_SOCKET, TIMEOUT, VIRSH_URI, VM_NAME};
+
+// to init env logger
+static INIT: Once = Once::new();
+
+fn run_test<T>(test: T) -> ()
+where
+    T: Send + 'static,
+    T: FnOnce() -> (),
+{
+    // init env_logger if necessary
+    INIT.call_once(|| {
+        env_logger::builder().is_test(true).init();
+    });
+    // setup before test
+    setup_test();
+
+    // setup test execution in a thread
+    let (done_tx, done_rx) = mpsc::channel();
+    let handle = thread::spawn(move || {
+        let val = test();
```

Comment:
What happens if a test assert fails? Won't this simply panic and prevent you from executing your teardown? Wouldn't you need something like `panic::catch_unwind()` in order to prevent that?

Reply:
https://medium.com/@ericdreichert/test-setup-and-teardown-in-rust-without-a-framework-ba32d97aa5ab

Reply:
`thread_spawn` will catch anything panicking in the thread's execution.
The result can be obtained when you call join():
https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#method.join

and the teardown code is still executed
example with a failing test:
~~~rust
    fn test_init_driver() {
        run_test(|| {
            assert!(false);
            init_driver();
        })
    }
~~~

`RUST_LOG=debug cargo test --features kvm test_init_driver -- --nocapture`
output
~~~
running 1 test
[2021-05-25T06:27:02Z DEBUG kvm] setup test
thread '<unnamed>' panicked at 'assertion failed: false', tests/kvm.rs:102:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2021-05-25T06:27:03Z DEBUG kvm] teardown test
thread 'tests::test_init_driver' panicked at 'Test timeout: Disconnected', tests/kvm.rs:43:9
test tests::test_init_driver ... FAILED
~~~